### PR TITLE
JDK-8281003 - MethodHandles::lookup throws NPE if caller is null

### PR DESCRIPTION
**Pull request by tprinzing**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8281003](https://bugs.openjdk.java.net/browse/JDK-8281003): MethodHandles::lookup throws NPE if caller is null
 * [JDK-8281765](https://bugs.openjdk.java.net/browse/JDK-8281765): MethodHandles::lookup throws NPE if caller is null (**CSR**)

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to 1bb6316f47bb4de03a0937510c1b18cbb63f7d90
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 7d1b97fe6a152a025d7ae7c2c8f333e5029026a3
 * [John R Rose](https://openjdk.java.net/census#jrose) (@rose00 - **Reviewer**) ⚠️ Review applies to 1bb6316f47bb4de03a0937510c1b18cbb63f7d90
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7447/head:pull/7447` \
`$ git checkout pull/7447`

Update a local copy of the PR: \
`$ git checkout pull/7447` \
`$ git pull https://git.openjdk.java.net/jdk pull/7447/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7447`

View PR using the GUI difftool: \
`$ git pr show -t 7447`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7447.diff">https://git.openjdk.java.net/jdk/pull/7447.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -116,7 +116,11 @@ private MethodHandles() { }  // do not instantiate
     @CallerSensitive
     @ForceInline // to ensure Reflection.getCallerClass optimization
     public static Lookup lookup() {
-        return new Lookup(Reflection.getCallerClass());
+        Class<?> c = Reflection.getCallerClass();
+        if (c == null) {
+            throw new IllegalCallerException();
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:121]: Throwing ICE is probably okay here, I just wonder if there is any practical advantage to having it return publicLookup instead, e.g. is there any scenario where a JNI attached thread might want to invoke a method with a Lookup parameter?

**Review by magicus — APPROVED** [review]
Build changes are OK

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 1 above)
Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:121]: `MethodHandles::publicLookup` can be called instead to get a public Lookup to invoke a method with a Lookup parameter.   The dilemma here is whether the API should be made null-caller friendly or using a proper API `MethodHandles::publicLookup` for such case.

File: test/jdk/java/lang/invoke/MethodHandles/exeNullCallerMethodHandlesLookup/NullCallerMethodHandlesLookupTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2019, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (AlanBateman) [inline test/jdk/java/lang/invoke/MethodHandles/exeNullCallerMethodHandlesLookup/NullCallerMethodHandlesLookupTest.java:2]: As this is a new file then I assume it should be 2022 only.

File: test/jdk/java/lang/invoke/MethodHandles/exeNullCallerMethodHandlesLookup/NullCallerMethodHandlesLookupTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import jdk.test.lib.Platform;
+import jdk.test.lib.process.OutputAnalyzer;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class NullCallerMethodHandlesLookupTest {
+    public static void main(String[] args) throws IOException {
+        Path launcher = Paths.get(System.getProperty("test.nativepath"), "NullCallerMethodHandlesLookupTest");
```
Comment (AlanBateman) [inline test/jdk/java/lang/invoke/MethodHandles/exeNullCallerMethodHandlesLookup/NullCallerMethodHandlesLookupTest.java:48]: Path.of might be nicer here but doesn't matter.

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 1 above)
Comment (mlchung) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:121]:
```suggestion
            throw new IllegalCallerException("no caller frame");
```

Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:121]: The javadoc needs to be updated to specify this `IllegalCallerException` be thrown.

```
    * @throws IllegalCallerException if there is no caller frame on the stack when called
    * directly from a JNI attached thread
```

**Review by mlchung — COMMENTED** [review]
This needs a CSR and the spec needs update.  

The test name could be shortened to `s/exeNullCallerMethodHandlesLookup/exeNullCallerLookup/`.

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 1 above)
Comment (rose00) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:121]: Agreed.  In this case there is no caller and any kind of fail-over to a designated caller would risk privilege escalation.  So we should throw.  I have no objection to throwing something more "diagnostic" than a NPE.

Arguably, JNI code is full-privileged, so someone might suggest, "just return a fully privileged lookup on some designated class".  But, even if such a class could be designated somehow (e.g., by rummaging down the  stack), handing out privileges on that class might be unexpected to the JNI author.  In fact, if the JNI code is working on behalf of a *low-privileged class* (whatever that means in context), then handing back a `Lookup` with higher privileges potentially leaks those privileges to the low-privileged class (depending on data flow, of course).  Trying to guess at a `Lookup` in this case would only create potential privilege escalations.  So we throw, and require the JNI programmer to say something clearer about their intentions.

**Review by rose00 — APPROVED** [review]

**Review by mlchung — APPROVED** [review]
Looks good.

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**